File: Sources/SwiftMTH/TelnetConstants.swift
// MARK: - Telnet Protocol Constants

/// Telnet command bytes.
public enum TelnetCommand {
    public static let IAC: UInt8   = 255
    public static let DONT: UInt8  = 254
    public static let DO: UInt8    = 253
    public static let WONT: UInt8  = 252
    public static let WILL: UInt8  = 251
    public static let SB: UInt8    = 250
    public static let GA: UInt8    = 249
    public static let EL: UInt8    = 248
    public static let EC: UInt8    = 247
    public static let AYT: UInt8   = 246
    public static let AO: UInt8    = 245
    public static let IP: UInt8    = 244
    public static let BREAK: UInt8 = 243
    public static let DM: UInt8    = 242
    public static let NOP: UInt8   = 241
    public static let SE: UInt8    = 240
    public static let EOR: UInt8   = 239
    public static let ABORT: UInt8 = 238
    public static let SUSP: UInt8  = 237
    public static let xEOF: UInt8  = 236

    /// True if `c` is a valid telnet command (>= xEOF).
    public static func isCommand(_ c: UInt8) -> Bool {
        c >= xEOF
    }
}

/// Telnet option numbers.
public enum TelnetOption {
    public static let ECHO: UInt8         = 1
    public static let SGA: UInt8          = 3
    public static let TTYPE: UInt8        = 24
    public static let EOR: UInt8          = 25
    public static let NAWS: UInt8         = 31
    public static let NEW_ENVIRON: UInt8  = 39
    public static let CHARSET: UInt8      = 42
    public static let MSDP: UInt8         = 69
    public static let MSSP: UInt8         = 70
    public static let MCCP2: UInt8        = 86
    public static let MCCP3: UInt8        = 87
    public static let MSP: UInt8          = 90
    public static let MXP: UInt8          = 91
    public static let GMCP: UInt8         = 201
}

/// Sub-constants used within subnegotiations.
public enum TelnetSub {
    // NEW-ENVIRON
    public static let ENV_IS: UInt8   = 0
    public static let ENV_SEND: UInt8 = 1
    public static let ENV_INFO: UInt8 = 2
    public static let ENV_VAR: UInt8  = 0
    public static let ENV_VAL: UInt8  = 1
    public static let ENV_ESC: UInt8  = 2
    public static let ENV_USR: UInt8  = 3

    // CHARSET
    public static let CHARSET_REQUEST: UInt8  = 1
    public static let CHARSET_ACCEPTED: UInt8 = 2
    public static let CHARSET_REJECTED: UInt8 = 3

    // MSSP
    public static let MSSP_VAR: UInt8 = 1
    public static let MSSP_VAL: UInt8 = 2
}

/// Announcement flags for the telnet option table.
public struct AnnounceFlags: OptionSet, Sendable {
    public let rawValue: Int
    public init(rawValue: Int) { self.rawValue = rawValue }

    public static let will = AnnounceFlags(rawValue: 1 << 0)
    public static let `do`   = AnnounceFlags(rawValue: 1 << 1)
}

/// Entry in the telnet option announcement table.
/// Determines which options are announced on connection.
public struct TelnetOptionEntry: Sendable {
    public let name: String
    public let announce: AnnounceFlags

    public init(_ name: String, _ announce: AnnounceFlags = []) {
        self.name = name
        self.announce = announce
    }
}

/// Default telnet option announcement table.
/// Index by telnet option number (0-255). Options with non-empty announce
/// flags will be sent as WILL/DO on connection.
public let defaultTelnetTable: [TelnetOptionEntry] = {
    var table = [TelnetOptionEntry](repeating: TelnetOptionEntry("", []), count: 256)

    table[1]   = TelnetOptionEntry("ECHO")
    table[3]   = TelnetOptionEntry("SUPPRESS GA")
    table[24]  = TelnetOptionEntry("TERMINAL TYPE", .do)
    table[25]  = TelnetOptionEntry("EOR")
    table[31]  = TelnetOptionEntry("NAWS", .do)
    table[39]  = TelnetOptionEntry("NEW_ENVIRON", .do)
    table[42]  = TelnetOptionEntry("CHARSET", .will)
    table[69]  = TelnetOptionEntry("MSDP", .will)
    table[70]  = TelnetOptionEntry("MSSP", .will)
    table[86]  = TelnetOptionEntry("MCCP2", .will)
    table[87]  = TelnetOptionEntry("MCCP3", .will)
    table[90]  = TelnetOptionEntry("MSP", .will)
    table[91]  = TelnetOptionEntry("MXP")
    table[201] = TelnetOptionEntry("GMCP", .will)

    return table
}()

File: kotlin/mth-core/src/main/kotlin/mth/core/TelnetConstants.kt
package mth.core

object TelnetCommand {
    const val IAC: Byte = 0xFF.toByte()  // 255
    const val DONT: Byte = 0xFE.toByte() // 254
    const val DO: Byte = 0xFD.toByte()   // 253
    const val WONT: Byte = 0xFC.toByte() // 252
    const val WILL: Byte = 0xFB.toByte() // 251
    const val SB: Byte = 0xFA.toByte()   // 250
    const val GA: Byte = 0xF9.toByte()   // 249
    const val EL: Byte = 0xF8.toByte()   // 248
    const val EC: Byte = 0xF7.toByte()   // 247
    const val AYT: Byte = 0xF6.toByte()  // 246
    const val AO: Byte = 0xF5.toByte()   // 245
    const val IP: Byte = 0xF4.toByte()   // 244
    const val BREAK: Byte = 0xF3.toByte() // 243
    const val DM: Byte = 0xF2.toByte()   // 242
    const val NOP: Byte = 0xF1.toByte()  // 241
    const val SE: Byte = 0xF0.toByte()   // 240
    const val EOR: Byte = 0xEF.toByte()  // 239
    const val ABORT: Byte = 0xEE.toByte() // 238
    const val SUSP: Byte = 0xED.toByte()  // 237
    const val xEOF: Byte = 0xEC.toByte()  // 236

    fun isCommand(c: Byte): Boolean = (c.toInt() and 0xFF) >= (xEOF.toInt() and 0xFF)
}

object TelnetOption {
    const val ECHO: Byte = 1
    const val SGA: Byte = 3
    const val TTYPE: Byte = 24
    const val EOR: Byte = 25
    const val NAWS: Byte = 31
    const val NEW_ENVIRON: Byte = 39
    const val CHARSET: Byte = 42
    const val MSDP: Byte = 69
    const val MSSP: Byte = 70
    const val MCCP1: Byte = 85.toByte()
    const val MCCP2: Byte = 86.toByte()
    const val MCCP3: Byte = 87.toByte()
    const val MSP: Byte = 90.toByte()
    const val MXP: Byte = 91.toByte()
    const val GMCP: Byte = 0xC9.toByte() // 201
}

object TelnetSub {
    const val ENV_IS: Byte = 0
    const val ENV_SEND: Byte = 1
    const val ENV_INFO: Byte = 2
    const val ENV_VAR: Byte = 0
    const val ENV_VAL: Byte = 1
    const val ENV_ESC: Byte = 2
    const val ENV_USR: Byte = 3

    const val CHARSET_REQUEST: Byte = 1
    const val CHARSET_ACCEPTED: Byte = 2
    const val CHARSET_REJECTED: Byte = 3

    const val MSSP_VAR: Byte = 1
    const val MSSP_VAL: Byte = 2
}
